adr/ADR-11.md

```diff
@@ -0,0 +1,271 @@
+
+# Next fetch request is a long fetch request with client side timeout
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.NQ2uAIXd4GoozKOTfECtIg 32
+{"expires":4990000000,"batch":5}
+```
+
+As part of the request payload, the `batch` field is set to the number
+of expected messages and `expires` is set to cancel the request
+`100ms` before the client side timeout.  In the example above, the
+timeout is `5s` so in the payload `expires` is the result of `5s -
+100ms` represented in nanoseconds.
+
+After making the first request no wait request, it is also recommended
+to send an auto unsubscribe protocol discounting the message already
+received as a result of the no wait request.  In case the batch
+request was for 5 messages, the client would auto unsubscribe after
+receiving 4:
+
+```shell
+UNSUB 1 4
+```
+
+When successful, the result of a batch request will be at least one
+message being delivered to the client.  In case not all messages are
+delivered and the client times out or goes away during the fetch
+request, it is recommended to unsubscribe to remove the interest of
+the awaited messages from the server, otherwise this risks the server
+sending messages to an inbox from a connected client that is no longer
+expecting messages.
+
+```shell
+SUB _INBOX.WvaJLnIXcj8Zf5SrxlHMTS  1
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 26
+{"batch":5,"no_wait":true}
+HMSG _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 8  28 28
+NATS/1.0 404 No Messages
+UNSUB 1 4
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 32
+{"expires":4990000000,"batch":5}
+MSG hello 1 $JS.ACK.bar.dur.2.29034.29041.1626845015078897000.0 4
+ping
+# Only 1 out of 5 messages receives, so client goes away and unsubscribes.
+UNSUB 1
+```
+
+##### Errors and Status messages handling
+
+While receiving messages, at any point the client may instead
+receive an error from the server as a status message.
+
+```sh
+MSG hello 1 $JS.ACK.bar.dur.2.29034.29041.1626845015078897000.0 5
+hello
+# (System reached too many inflight fetch requests condition)
+HMSG _INBOX.WvaJLnIXcj8Zf5SrxlHMRI 1  32 32
+NATS/1.0 408 Request Timeout
+```
+
+Whenever there is a status message, it is recommended that the error
+is not returned to the user as a processable message and instead handle
+is as the error condition:
+
+```go
+for {
+	msgs, err := sub.Fetch(5)
+	if err != nil {
+		// Any error such as:
+		// 
+		// - client timeout
+		// - request timeout
+		// - bad request
+		// - no responders
+		// - system unavailable (no current JS quorum)
+		log.Println("Error:", err)
+		continue
+	}
+	// Msgs here are never status or error messages,
+	// in case there is an error the loop just breaks.
+	for _, msg := range msgs {
+		msg.Ack()
+	}
+}
+```
+
+#### Pull(1)
+
+For the case of fetching a single message, it is possible to optimize
+things by making the first fetch request as a no wait request instead
+and by preparing a new style like request/response handler.
```

Comment:
```suggestion
and by preparing a new style like request/response handler using a wildcard
subscription.  This will result in less chatty protocol and also works better with
topologies where JetStream is running as a leafnode on the edge.
```